crates/rover-client/src/introspection/schema.rs

```diff
@@ -0,0 +1,237 @@
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Schema {
+    types: Vec<SchemaTypes>,
+    directives: Vec<SchemaDirectives>,
+}
+
+impl Schema {
+    /// Encode Schema into an SDL.
+    pub fn encode(self) -> String {
+        let mut sdl = SDL::new();
+
+        // Exclude GraphQL directives like 'skip' and 'include' before encoding directives.
+        self.directives
+            .into_iter()
+            .filter(|directive| !GRAPHQL_DIRECTIVES.contains(&directive.name.as_str()))
+            .for_each(|directive| Self::encode_directives(directive, &mut sdl));
+
+        // Exclude GraphQL named types like __Schema before encoding full type.
+        self.types
+            .into_iter()
+            .filter(|type_| match type_.full_type.name.as_deref() {
+                Some(name) => !GRAPHQL_NAMED_TYPES.contains(&name),
+                None => false,
+            })
+            .for_each(|type_| Self::encode_full_type(type_, &mut sdl));
+
+        sdl.finish()
+    }
+
+    fn encode_directives(directive: SchemaDirectives, sdl: &mut SDL) {
+        let mut directive_ = Directive::new(directive.name);
+        directive_.description(directive.description);
+        for location in directive.locations {
+            // Location is of a __DirectiveLocation enum that doesn't implement
+            // Display (meaning we can't just do .to_string). This next line
+            // just forces it into a String with format! debug.
+            directive_.location(format!("{:?}", location));
+        }
+
+        sdl.directive(directive_)
+    }
+
+    fn encode_full_type(type_: SchemaTypes, sdl: &mut SDL) {
+        let ty = type_.full_type;
+
+        match ty.kind {
+            __TypeKind::OBJECT => {
+                let mut object_def = ObjectDef::new(ty.name.unwrap_or_else(String::new));
+                object_def.description(ty.description);
+                if let Some(interfaces) = ty.interfaces {
+                    for interface in interfaces {
+                        object_def.interface(interface.type_ref.name.unwrap_or_else(String::new));
+                    }
+                }
+                if let Some(field) = ty.fields {
+                    for f in field {
+                        let field_def = Self::encode_field(f);
+                        object_def.field(field_def);
+                    }
+                    sdl.object(object_def);
+                }
+            }
+            __TypeKind::INPUT_OBJECT => {
+                let mut input_def = InputObjectDef::new(ty.name.unwrap_or_else(String::new));
+                input_def.description(ty.description);
+                if let Some(interfaces) = ty.interfaces {
+                    for interface in interfaces {
+                        input_def.interface(interface.type_ref.name.unwrap_or_else(String::new));
+                    }
+                }
+                if let Some(field) = ty.fields {
```

Comment:
I think this should be `ty.input_fields`, since `fields` and `inputFields` are separate fields on `__Type`?

Reply:
Ohhhhh, I've been thinking this whole time that I only get `fields` for this kind. The compiler never complained about the field `fields` not being present on `ty`, which I guess means that I get both.

Reply:
Yeah, in the introspection query most fields on `__Type` are optional since it describes all possible kinds of types.

Reply:
It makes more sense to me now to have InputField and regular fields separated, like you've mentioned several times. It was weird to have essentially duplicate code to generate the exact same structure for the exact same return value of `ty.fields` before. Which is why I said that would make a lot more sense once we have an AST to walk and avoid having to write out the exact same code over the exact same set of values. 

I will still not be able to have both `InputObject` use `InputValue` since the formatting for an argument is different than a field:
```js
//  input values are single lined at the moment
name: String, val: Int, val2: Int
```
```js
// we will call these input fields so we can multiline them and format them similarly to fields
name: String
val: Int
val2: Int
```

It _is_ a bit weird to me that graphql spec does not differentiate between an input field and an input value, as they do format differently, but I guess that's not something to worry about atm